packages/gatsby/src/internal-plugins/functions/gatsby-node.ts

```diff
@@ -0,0 +1,318 @@
+import fs from "fs-extra"
+import glob from "glob"
+import path from "path"
+import webpack from "webpack"
+import multer from "multer"
+import * as express from "express"
+import { urlResolve } from "gatsby-core-utils"
+import { ParentSpanPluginArgs, CreateDevServerArgs } from "gatsby"
+import TerserPlugin from "terser-webpack-plugin"
+import { internalActions } from "../../redux/actions"
+import { reportWebpackWarnings } from "../../utils/webpack-error-utils"
+import formatWebpackMessages from "react-dev-utils/formatWebpackMessages"
+import dotenv from "dotenv"
+import chokidar from "chokidar"
+
+const isProductionEnv = process.env.gatsby_executing_command !== `develop`
+
+const createWebpackConfig = async ({
+  siteDirectoryPath,
+  functionsDirectory,
+  store,
+}): webpack.Configuration => {
+  const files = await new Promise((resolve, reject) => {
+    const functionsGlob = `**/*.{js,ts}`
+    glob(functionsGlob, { cwd: functionsDirectory }, (err, files) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve(files)
+      }
+    })
+  })
+
+  if (files?.length === 0) {
+    reporter.warn(
+      `No functions found in directory: ${path.relative(
+        siteDirectoryPath,
+        functionsDirectory
+      )}`
+    )
+  }
+
+  const knownFunctions = new Map(
+    files.map(file => [
+      urlResolve(path.parse(file).dir, path.parse(file).name),
+      file,
+    ])
+  )
+
+  store.dispatch(internalActions.setFunctions(knownFunctions))
+
+  // Load environment variables from process.env.GATSBY_* and .env.* files.
+  // Logic is shared with webpack.config.js
+
+  // node env should be DEVELOPMENT | PRODUCTION as these are commonly used in node land
+  const nodeEnv = process.env.NODE_ENV || `${defaultNodeEnv}`
+  // config env is dependent on the env that it's run, this can be anything from staging-production
+  // this allows you to set use different .env environments or conditions in gatsby files
+  const configEnv = process.env.GATSBY_ACTIVE_ENV || nodeEnv
+  const envFile = path.join(siteDirectoryPath, `./.env.${configEnv}`)
```

Comment:
People not always use `.env.development` / `.env.production`, they also use a `.env` for everything. Needs to be accounted for here

Reply:
Is there a reason we can't just take this from `process.env`, so users can decide for themselves how to pass them in?

Reply:
We do also take from `process.env` here.

The reason I didn't add support for `.env` is we don't do that for client side code — https://www.gatsbyjs.com/docs/how-to/local-development/environment-variables#defining-environment-variables

So it could be weird if people adopt `.env` as their convention and then variables there don't show up in the client.

I agree with you that `.env` should be the default and perhaps the `.env.*` just override the defaults but that change would need made across everywhere.

Reply:
https://github.com/gatsbyjs/gatsby/blob/4a9fffd1e3236e335bffcfa3a60c30a3cc15dca2/packages/gatsby/src/utils/webpack.config.js#L63-L74

Reply:
This was something people asked for in the v3 wishlist as well. Perhaps we can add it behind a flag now & enabled it in v4.

Reply:
Oh and right @ascorbic, if they load a .env file in their gatsby-node.js file, that'd get picked up here. One downside is it changes to it wouldn't get hot reloaded like they are for .env.* but that's fine.

Reply:
Ideally people wouldn't need to do the dotenv thing themselves etc. -- but that is tracked here: https://github.com/gatsbyjs/gatsby/discussions/30893

Reply:
👍